### ISSUE TITLE
Port StyleCop Documentation Rules (but just the useful ones)

### ISSUE THREAD

**Issue by bricelam** [dotnet/roslyn-analyzers#779]
I've always found the [StyleCop Documentation Rules](http://www.stylecop.com/docs/Documentation%20Rules.html) that check basic correctness useful. Are there any plans to port these? Here are the particular ones I would like to see.

Rule | Description
--- | ---
SA1603 | The XML is well-formed
SA1606<br />SA1607<br />SA1610<br />SA1613<br />SA1614<br />SA1616<br />SA1621<br />SA1622 | Elements and attributes aren't empty
SA1612<br />SA1620 | Referenced parameters exist
SA1617 | No `<return>` on `void` methods
SA1624 | No *Gets __or sets__...* on read-only properties

**Comment by srivatsn** [comment]
The stylecop rules are being implemented as analyzer in the https://github.com/dotnetanalyzers/stylecopanalyzers project. Most of the rules you mention above seem to have already been implemented according to [their status page](https://stylecop.pdelvo.com/). You can get these analyzers through the [StyleCop.Analyzers NuGet Package](http://www.nuget.org/packages/StyleCop.Analyzers/1.0.0-rc3)

**Comment by lgolding** [comment]
Thanks for filing the issue, @bricelam. I am not sure how valuable those rules are, since it's so easy for tools or humans to generate valid but useless doc comments. We've all see this sort of thing:

```C#
/// <summary>
/// Gets a Foo
/// </summary>
/// <param name="type">
/// A type.
/// </param>
/// <returns>
/// A foo.
/// </returns>
public Foo GetFoo(string type)
```

But we will definitely consider these.

**Comment by lgolding** [comment]
... oh. I see Sri commented as well.

**Comment by michaelcfanning** [comment]
The compiler apparently already provides placeholders for built-in analysis for doc comments, including some of what's requested here (badly formed XML is CS1570). You must enable doc XML generation in the project build to see the results. My suggestion for this area has been that we should either build out this capability (within VS) or we should break out all the doc comment analysis into its own analyzer.

![doccomments](https://cloud.githubusercontent.com/assets/13526455/12177667/2e1ff10c-b522-11e5-8b93-187b364c1f56.PNG)

**Comment by lgolding** [comment]
Since doc comments are part of the C# language, it makes sense for the well-formedness checks to be done by the compiler. Checks such as spell-checking the comments could go in Text.Analyzers. But if there are any other checks (for example, complaining about vacuous comments such as "`<param name="type">A type</param>`," then I agree we should create an XmlDocComments.Analyzers package for them. Then we'd have to decide whether the spell-checking analyzer belongs in Text.Analyzers or XmlDocComments.Analyzers.

**Comment by sharwell** [comment]
@bricelam All of the items you mentioned have been implemented in StyleCop Analyzers. We don't report SA1603 directly, but as mentioned in DotNetAnalyzers/StyleCopAnalyzers#1291, there is no need for this warning since it is an exact duplicate of CS1570.

**Comment by bricelam** [comment]
Groovy, thanks. Between all the different projects, it sounds like I have what I need.

From my perspective, this issue can be closed. But there does seem to still be some unresolved discussion...

@michaelcfanning said:

> My suggestion for this area has been that we should either build out this capability (within VS) or we should break out all the doc comment analysis into its own analyzer.

@lgolding said:

> But if there are any other checks (for example, complaining about vacuous comments such as `<param name="type">A type</param>`), then I agree we should create an XmlDocComments.Analyzers package for them.

**Comment by michaelcfanning** [comment]
Agreed, there's still some useful discussion here. Factoring checks is an imperfect science. :) One thing I believe, though, is that where a body of analysis coheres, it's best for developers if that topic is covered in a single place. This is the rationale behind grouping the .NET design guidelines in a common place, you care about your API surface area or you don't. I think XML documentation is similar and so I'd prefer to see a single body of analysis for this topic that we could all contribute to. Otherwise the story is, enable these built-in checks, install StyleCop and enable these other checks, etc. etc.

When you consider a cross-cutting analysis capability such as spell-check, things can get muddied. Also, it's tempting (and appropriate) to think about different factorings for analysis that entails a different level of developer involvement (correcting a typo is a very different exercise than authoring a meaningful missing doc comment).

My opinion is that we should think about rules factoring in terms of the value that's being offered to customers by the developer running the analysis ('I want to developer the best reusable components for my customer audience, so I run the .NET design guideline checks', 'I want to deliver the most thorough and correct XML doc comments supporting my component', etc.).

**Comment by sharwell** [comment]
It would be interesting to separate what is currently StyleCop Analyzers into multiple analyzer packages:

1. Spacing, layout, and layout-related readability rules
1. Remaining readability rules and maintainability rules
1. Naming rules
1. Ordering rules
1. Documentation rules

If we did this, we could still have the StyleCop.Analyzers package in NuGet, but it would be a metadata-only package with dependencies declared for each of the others.

**Comment by lgolding** [comment]
@sharwell That's a nice idea. @michaelcfanning, @srivatsn what do you think?

**Comment by lgolding** [comment]
I'd like to close this issue, as @bricelam 's original concern ("where can I find XML doc comment analyzers?") has been addressed. We certainly can and should continue to discuss and refine our plans for factoring analyzers into coherent packages.

Specifically with regard to the StyleCop analyzers, in the near term it's up to @sharwell and his team what they want to do. As we have discussed, one of our goals is to make a comprehensive, configurable set of style analyzers available out of the box with VS (see the "Analyzer ecosystem [Inventory of Epics](https://github.com/dotnet/roslyn-analyzers-contrib/blob/master/docs/Inventory-of-Epics.md) document). As we move closer to that goal, we'll want to align on our vision of how to factor the style analyzers. But I do think we're basically in agreement that small, focused packages of analyzers are preferable to large monolithic packages, so I expect we'll easily come to agreement on the details.
